_drafts/fs2.md

```diff
@@ -0,0 +1,722 @@
+
+The fs2 library implements the bracket pattern through the `bracket` method:
+
+```scala
+// fs2 library code
+def bracket[F[x] >: Pure[x], R](acquire: F[R])(release: R => F[Unit]): Stream[F, R] = ???
+```
+
+The function `acquire` is used to acquire a resource, and the function `release` is used to release the resource. The resulting stream has elements of type `R`. Moreover, **the resource is released at the end, both in case of success and in case of error**. Notice that both the `acquire` and `release` functions returns an effect since they can throw exceptions during the acquisition or release of resources.
+
+Let's make an example. We want to use a resource during the persistence of the stream containing the actors of the JLA. First, we define a value class representing a connection to a database:
+
+```scala
+case class DatabaseConnection(connection: String) extends AnyVal
+```
+
+We will use the `DatabaseConnection` as the resource we want to acquire and release through the bracket pattern. Then, the acquiring and releasing function:
+
+```scala
+val acquire = IO {
+  val conn = DatabaseConnection("jlaConnection")
+  println(s"Acquiring connection to the database: $conn")
+  conn
+}
+val release = (conn: DatabaseConnection) => 
+  IO.println(s"Releasing connection to the database: $conn")
+```
+
+Finally, we use them to call the `bracket` method and then save the actors in the stream:
+
+```scala
+val managedJlActors: Stream[IO, Int] = 
+  Stream.bracket(acquire)(release).flatMap(conn => savedJlActors)
+```
+
+Since the `savedJlActors` we defined some time ago throws an exception when the stream is evaluated, the `managedJlActors` stream will terminate with an error. The `release` function is called, and the `conn` value is released. The execution output of the `managedJlActors` stream should be something similar to the following:
+
+```
+Acquiring connection to the database: DatabaseConnection(jlaConnection)
+Saving actor: Actor(0,Henry,Cavill)
+Releasing connection to the database: DatabaseConnection(jlaConnection)
+java.lang.RuntimeException: Something went wrong during the communication with the persistence layer
+```
+
+As we can see, the resource we created was released, even if the stream had terminated with an error. That's awesome!
+
+## 6. The `Pull` Type
+
+As we said more than once, the fs2 defines streams as a _pull_ type, which means that the **stream effectively computes the next stream element just in time**.
+
+Under the hood, the library implements the `Stream` type functions using the `Pull` type to honor this behavior. This type, also available as a public API, lets us implement streams using the pull model.
+
+The `Pull[F[_], O, R]` type represents a program that can pull output values of type `O` while computing a result of type `R` while using an effect of type `F`. As we can see, the type introduces the new type variable `R` that is not available in the `Stream` type.
```

Comment:
I'd rephrase this a bit.

As I see it, Pull is a low-level construct that allows us to manually manage an element of type O from a stream, under an effect type F, resulting in a type R which we can then recursively use to manage the next elements in the stream.

With Pull, we can describe in minute detail _how_ elements in a stream are produced.

Reply:
Pull is very powerful, but too low-level for the video version of this tutorial. Maybe I'll show Pull separately